packages/web-components/fast-components/src/design-system-provider/README.md

```diff
@@ -0,0 +1,260 @@
+    <p style="background: var(--background-color)">#FFFFFF</p>
+
+    <fast-design-system-provider design-unit="8">
+        <p style="height: calc(var(--design-unit) * 1px)">8px</p>
+        <p style="background: var(--background-color)">#FFFFFF</p>
+    </fast-design-system-provider>
+</fast-design-system-provider>
+```
+
+
+### Recipes
+
+There are some things that CSS just can't do. Advanced math, array manipulation, and conditionals are simply not possible with today's CSS.
+
+To address this gap, the _DesignSystemProvider_ is capable of registering `CSSCustomPropertyDefinition` types through the `registerCSSCustomProperty` method:
+
+**EXAMPLE: Register a CSS custom property that is a function of the _design system_**
+
+```ts
+DesignSystemProvider.registerCSSCustomProperty({
+    name: "design-unit-12th",
+    value: designSystem => Math.pow(designSystem.designUnit, 12),
+});
+```
+
+In the above example, the value function will be re-evaluated if the _design system_ ever changes.
+
+The above API is made especially useful when defining a [component stylesheet](https://github.com/microsoft/fast-dna/blob/master/packages/web-components/fast-element/docs/building-components.md#defining-css) - individual stylesheets can declare dependencies on CSS custom properties that are functions of the element instance's *design system*
+
+**EXAMPLE: creating a recipe dependency**
+```ts
+import { css } from "@microsoft/fast-element";
+import { cssCustomPropertyBehaviorFactory, DesignSystemProvider } from "@microsoft/fast-components";
+
+const styles = css`
+    :host {
+        height: var(--fancy-height);
+    }
+`.withBehaviors(
+    cssCustomPropertyBehaviorFactory(
+        "fancy-height",
+        designSystem => Math.pow(designSystem.designUnit, 12),
+        DesignSystemProvider.findProvider
+    )
+)
+```
+
+### Creating a Design System Provider
+
+To create a new _Design System Provider_, extend the `DesignSystemProvider` class and decorate it with the `@designSystemProvider` decorator, providing the decorator the element _tag name_ you wish to use:
+
+**EXAMPLE: Creating a custom Design System Provider**
+
+```ts
+import { DesignSystemProvider, designSystemProvider } from "@microsoft/fast-components";
+
+@designSystemProvider("fancy-design-system-provider")
+class FancyDesignSystemProvider extends DesignSystemProvider {}
+```
+
+```html
+<fancy-design-system-provider use-defaults>
+    Hello Fancy World!
+</fancy-design-system-provider>
+```
+
+#### Declaring Design System Properties
+
+Building off the above, _design system_ properties can be declared using the `@designSystemProperty` decorator.
+
+**EXAMPLE: Creating a design system property declaration**
+
+```ts
+// ..
+@designSystemProvider("fancy-design-system-provider")
+class FancyDesignSystemProvider extends DesignSystemProvider {
+    @attr("fancy-property")
+    @designSystemProperty({ cssCustomProperty: "fancy-property", default: "red" })
```

Comment:
Again, not a critique of the docs, but it would be cool if the `designSystemProperty` could automatically pick up the `attr` name so that it didn't need to be provided (unless is was different).

Reply:
Good idea - I'll write up a proposal for that.